# Contents

### 01. Import Libraries and Data
### 02. Data Exploration
### 03. Data Wrangling
### 04. Data Consistency Checks
    Mixed-Type Data
    Missing Data
    Duplicates
### 05. Combining customers with ords_prods
### 06. Exporting Data

# 01. Import Libraries and Data

In [2]:
import pandas as pd
import numpy as np
import os

In [3]:
# Assign path to str
path = r'/Users/allyfornino/Documents/10-2022 Instacart Basket Analysis/'

In [4]:
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

# 02. Data Exploration

In [5]:
customers.head(10)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374  
5             2    married   49643  
6             0     single   61746  
7             0     single   63712  
8             1    married  162432  
9             2    married   32072

In [6]:
customers.shape

(206209, 10)

In [7]:
customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

# 03. Data Wrangling

In [8]:
# Renaming columns
customers.rename(columns = {'First Name' : 'first_name', 'Surnam' : 'last_name', 'Gender' : 'gender', 'STATE' : 'state', 'Age' : 'age', 'n_dependants' : 'num_dependents', 'fam_status' : 'marital_status'}, inplace = True)

In [9]:
# Check output
customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_dependents marital_status  income  
0               3        married  165665  
1               0         single   59285  
2               2        married   99568  
3               0         single   42049  
4               1        married   40374

In [10]:
# Changing data type of user_id to str
customers['user_id'] = customers['user_id'].astype(str)

In [11]:
# Checking data type of user_id
customers['user_id'].dtype

dtype('O')

# 04. Data Consistency Checks

## Mixed-Type Data

In [12]:
# Checking for any mixed-type data
for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len(customers[weird]) > 0:
      print(col)

first_name


In [13]:
# Checking for missing values to see if that is the cause of mix
customers['first_name'].isnull().sum()

11259

In [14]:
customers[customers['first_name'].isnull() == True]

user_id first_name last_name  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        num_dependents marital_status  income  
53                   2        married   41709  
73                   0         single   82518  
82                   3        married  117099  
99                   1        married  155673  
105                  3        married  151819  
...                ...            ...     ...  
206038               3        married   87783  
206044               2        married  108488  
206090               1        married   45275  
206162               0         single   56325  
206171               2        married   59222  

[11259 rows x 10 columns]

#### There are 11259 missing values for first name. We are still able to identify customers with the user_id key so the missing values will remain NaN

In [15]:
# Making all values in first_name strings
customers['first_name'] = customers['first_name'].astype('str')

## Missing Data

In [16]:
# Checking for missing data in rest of df
customers.isnull().sum()

user_id           0
first_name        0
last_name         0
gender            0
state             0
age               0
date_joined       0
num_dependents    0
marital_status    0
income            0
dtype: int64

#### There is no other missing data

## Duplicates

In [17]:
# Checking for duplicates
customers_dups = customers[customers.duplicated()]

In [18]:
customers_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, num_dependents, marital_status, income]
Index: []

#### There are no duplicates

In [19]:
# Clean customers df before combining
customers

user_id first_name last_name  gender           state  age date_joined  \
0        26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1        33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2        65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3       125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4       130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...        ...        ...       ...     ...             ...  ...         ...   
206204  168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205   49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206  135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207   81095       Rose   Rollins  Female      California   27    4/1/2020   
206208   80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        num_dependents marital_status  income  
0                    3        married  165665  
1                    0         single   59285  
2                    2        married   99568  
3                    0         single   42049  
4                    1        married   40374  
...                ...            ...     ...  
206204               1        married  148828  
206205               3        married  168639  
206206               2        married   53374  
206207               1        married   99799  
206208               1        married   57095  

[206209 rows x 10 columns]

In [20]:
# Exporting customers as pkl
customers.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'customers_clean.pkl'))

# 05. Combining customers with ords_prods

In [21]:
# Import ords_prods_merge'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_updated2.pkl'))

In [22]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   

   reordered  ... prices  _merge     busiest_day  busiest_period_of_day  \
0          0  ...    9.0    both  Regularly busy         Average orders   
1          1  ...    9.0    both    Slowest days         Average orders   
2          1  ...    9.0    both    Slowest days            Most orders   
3          1  ...    9.0    both    Slowest days         Average orders   
4          1  ...    9.0    both    Slowest days            Most orders   

  max_order  loyalty_flag average_order  spending_flag order_frequency  \
0        10  New customer      6.367797    Low spender            20.5   
1        10  New customer      6.367797    Low spender            20.5   
2        10  New customer      6.367797    Low spender            20.5   
3        10  New customer      6.367797    Low spender            20.5   
4        10  New customer      6.367797    Low spender            20.5   

          frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer  

[5 rows x 23 columns]

In [23]:
# Dropping _merge column from previous merge
ords_prods_merge = ords_prods_merge.drop('_merge', axis = 1)

In [24]:
ords_prods_merge.shape

(32404859, 22)

In [25]:
# Checking datatype of dfs to ensure they match
ords_prods_merge.dtypes

order_id                    int64
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
new_customer                 bool
product_id                  int64
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
busiest_day                object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
average_order             float64
spending_flag              object
order_frequency           float64
frequency_flag             object
dtype: object

In [26]:
ords_prods_merge['user_id'] = ords_prods_merge['user_id'].astype('str')

In [27]:
# Testing merge
pd.merge(ords_prods_merge, customers, on = 'user_id', indicator = True)

order_id user_id  order_number  order_day_of_week  \
0          2539329       1             1                  2   
1          2398795       1             2                  3   
2           473747       1             3                  3   
3          2254736       1             4                  4   
4           431534       1             5                  4   
...            ...     ...           ...                ...   
32404854    156685  106143            26                  4   
32404855    484769   66343             1                  6   
32404856   1561557   66343             2                  1   
32404857    276317   66343             3                  6   
32404858   2922475   66343             4                  1   

          order_hour_of_day  days_since_prior_order  new_customer  product_id  \
0                         8                     NaN          True         196   
1                         7                    15.0         False         196   
2                        12                    21.0         False         196   
3                         7                    29.0         False         196   
4                        15                    28.0         False         196   
...                     ...                     ...           ...         ...   
32404854                 23                     5.0         False       19675   
32404855                 11                     NaN          True       47210   
32404856                 11                    30.0         False       47210   
32404857                 15                    19.0         False       47210   
32404858                 12                    30.0         False       47210   

          add_to_cart_order  reordered  ...  first_name  last_name  gender  \
0                         1          0  ...       Linda     Nguyen  Female   
1                         1          1  ...       Linda     Nguyen  Female   
2                         1          1  ...       Linda     Nguyen  Female   
3                         1          1  ...       Linda     Nguyen  Female   
4                         1          1  ...       Linda     Nguyen  Female   
...                     ...        ...  ...         ...        ...     ...   
32404854                  1          1  ...      Gerald      Yates    Male   
32404855                  1          0  ...  Jacqueline     Arroyo  Female   
32404856                  1          1  ...  Jacqueline     Arroyo  Female   
32404857                  1          1  ...  Jacqueline     Arroyo  Female   
32404858                  1          1  ...  Jacqueline     Arroyo  Female   

              state age date_joined  num_dependents marital_status  income  \
0           Alabama  31   2/17/2019               3        married   40423   
1           Alabama  31   2/17/2019               3        married   40423   
2           Alabama  31   2/17/2019               3        married   40423   
3           Alabama  31   2/17/2019               3        married   40423   
4           Alabama  31   2/17/2019               3        married   40423   
...             ...  ..         ...             ...            ...     ...   
32404854     Hawaii  25   5/26/2017               0         single   53755   
32404855  Tennessee  22   9/12/2017               3        married   46151   
32404856  Tennessee  22   9/12/2017               3        married   46151   
32404857  Tennessee  22   9/12/2017               3        married   46151   
32404858  Tennessee  22   9/12/2017               3        married   46151   

         _merge  
0          both  
1          both  
2          both  
3          both  
4          both  
...         ...  
32404854   both  
32404855   both  
32404856   both  
32404857   both  
32404858   both  

[32404859 rows x 32 columns]

In [28]:
ords_prods_custs = ords_prods_merge.merge(customers, on = 'user_id', indicator = True)

In [29]:
ords_prods_custs.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   

   reordered  ... first_name  last_name  gender    state age date_joined  \
0          0  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
1          1  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
2          1  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
3          1  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
4          1  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   

   num_dependents marital_status  income _merge  
0               3        married   40423   both  
1               3        married   40423   both  
2               3        married   40423   both  
3               3        married   40423   both  
4               3        married   40423   both  

[5 rows x 32 columns]

In [30]:
ords_prods_custs.shape

(32404859, 32)

# 06. Exporting Combined Data

In [31]:
ords_prods_custs.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_all.pkl'))